src/plugins/profiling/server/routes/search_topn.ts

```diff
@@ -6,11 +6,68 @@
  * Side Public License, v 1.
  */
 import { schema } from '@kbn/config-schema';
-import type { IRouter } from 'kibana/server';
-import { IEsSearchRequest } from '../../../data/server';
-import { IEsSearchResponse } from '../../../data/common';
+import type { IRouter, KibanaResponseFactory } from 'kibana/server';
+import {
+  AggregationsHistogramBucket,
+  AggregationsMultiBucketAggregateBase,
+} from '@elastic/elasticsearch/lib/api/types';
 import type { DataRequestHandlerContext } from '../../../data/server';
 import { getRemoteRoutePaths } from '../../common';
+import { projectTimeRangeQuery, autoHistogramSumCountOnGroupByField } from './mappings';
+
+export async function topNElasticSearchQuery(
+  context: DataRequestHandlerContext,
+  index: string,
+  projectID: string,
+  timeFrom: string,
+  timeTo: string,
+  searchField: string,
+  response: KibanaResponseFactory
+) {
+  const esClient = context.core.elasticsearch.client.asCurrentUser;
+  const resTopNStackTraces = await esClient.search(
+    {
+      index,
+      body: {
+        query: projectTimeRangeQuery(projectID, timeFrom, timeTo),
+        aggs: {
+          histogram: autoHistogramSumCountOnGroupByField(searchField),
+        },
+      },
+    },
+    {}
+  );
+
+  if (searchField === 'StackTraceID') {
+    const docIDs: string[] = [];
+    const autoDateHistogram = resTopNStackTraces.body.aggregations
+      ?.histogram as AggregationsMultiBucketAggregateBase<AggregationsHistogramBucket>;
+
+    autoDateHistogram.buckets?.forEach((timeInterval: any) => {
+      timeInterval.group_by.buckets.forEach((stackTraceItem: any) => {
+        docIDs.push(stackTraceItem.key);
+      });
+    });
+
+    const resTraceMetadata = await esClient.mget<any>({
+      index: 'profiling-stacktraces',
+      body: { ids: docIDs },
+    });
+
+    return response.ok({
+      body: {
+        topN: resTopNStackTraces.body.aggregations,
+        traceMetadata: resTraceMetadata.body.docs,
+      },
+    });
+  } else {
+    return response.ok({
+      body: {
+        topN: resTopNStackTraces.body.aggregations,
+      },
+    });
+  }
+}
```

Comment:
Right now I left this in the query execution, but it could be good to extract the response handling into its own function, to test it.
I'll probably add 1 more commit

Reply:
FYI I didn't do this yet... probably worth a dedicated PR since the size of this one is already relevant 😅